pkg/controller/master/image/backing_image_controller.go

```diff
@@ -46,11 +65,16 @@ func (h *backingImageHandler) OnChanged(_ string, backingImage *lhv1beta2.Backin
 			toUpdate = handleFail(toUpdate, condition.Cond(harvesterv1beta1.ImageImported), fmt.Errorf(status.Message))
 			toUpdate.Status.Progress = status.Progress
 		} else if status.State == lhv1beta2.BackingImageStateReady || status.State == lhv1beta2.BackingImageStateReadyForTransfer {
-			harvesterv1beta1.ImageImported.True(toUpdate)
-			harvesterv1beta1.ImageImported.Reason(toUpdate, "Imported")
-			harvesterv1beta1.ImageImported.Message(toUpdate, status.Message)
 			toUpdate.Status.Progress = status.Progress
 			toUpdate.Status.Size = backingImage.Status.Size
+			toUpdate.Status.VirtualSize = backingImage.Status.VirtualSize
```

Comment:
When upgrading an old version of Harvester, with existing images the image CRD will already be imported and this code will not be reached. Therefore the virtual image size will never be correctly displayed for existing image CRDs.

Reply:
The `if` on line 59 takes care of that. We only return up there if the image CRD is imported and the virtualSize is set. If the virtualSize isn't set (which it won't be for existing imported images), we land here.

Reply:
But the entire function will only be called when there is a change on the CRD, right? When an image is already imported, wouldn't the CRD always stay the same?

Reply:
Dammit.

This `OnChanged()` function is called when the longhorn backing image CR changes, which it will, when a version of Longhorn which supports backing image virtualSize is installed/starts up. So for existing images, what _should_ happen (and what _did_ happen when I was originally testing this during development) is something like this:

- Longhorn's backing-image-manager determines virtualSize of existing backing image
- Longhorn's longhorn-manager, in `syncBackingImageFileInfo()` picks up the virtualSize from the backing image, and sets that in the lhbi CR.
- Harvester notices that change, propagates the change to the existing imported vmimage CR and everyone is happy.

Since then though, the virtualSize addition to longhorn has made it into LH version 1.5.5 (which is included in Harvester 1.2.2) and LH 1.6.2 (included in Harvester 1.3.1). Which is why I started this comment with "dammit", because I fear you may be right -- if we're running Harvester 1.2.2 or 1.3.1, the lbhi CRs will have virtualSize set already so should have no reason to change.

I'm gonna need to re-test this...

Thanks for the review!

Reply:
OK, confirmed (dammit!)

Tested as follows:

- Install Harvester v1.3.0 (no virtualSize support anywhere).
- Add an image (https://download.opensuse.org/distribution/leap/15.5/appliances/openSUSE-Leap-15.5-Minimal-VM.x86_64-Cloud.qcow2)
- Check lhbi:
```
# kubectl get lhbi -A
NAMESPACE         NAME                  UUID       SOURCETYPE   SIZE        AGE
longhorn-system   default-image-kmvkf   006145f3   download     260465664   107s
```
- Upgrade to harvester v1.3.1 (LH supports virtualSize, Harvester does not yet)
- Check lhbi and vmimages:
```
# kubectl get lhbi -A
NAMESPACE         NAME                  UUID       SOURCETYPE   SIZE        VIRTUALSIZE   AGE
longhorn-system   default-image-kmvkf   006145f3   download     260465664   833617920     33m

# kubectl get vmimages
NAME          DISPLAY-NAME                                       SIZE        AGE
image-kmvkf   opensuse-leap-15.5-minimal-vm.x86_64-cloud.qcow2   260465664   58m
```
- Build a version of the v1.3.1 harvester container image with this PR applied, then run `kubectl -n harvester-system edit deployments/harvester` and replace `image: rancher/harvester:v1.3.1` with `image: tserong/harvester:wip-v1.3.1-virtualSize-head`. Also run `kubectl edit crd/virtualmachineimages.harvesterhci.io` to apply the CRD changes.
- Check vmimages and observe no virtualSize :-/ :
```
# kubectl get vmimages
NAME          DISPLAY-NAME                                       SIZE        VIRTUALSIZE   AGE
image-kmvkf   opensuse-leap-15.5-minimal-vm.x86_64-cloud.qcow2   260465664                 64m
```

If I subsequently add another image, that one of course works fine, but as you suspected the old one never gets updated:

```
# kubectl get lhbi -A
NAMESPACE         NAME                  UUID       SOURCETYPE   SIZE        VIRTUALSIZE   AGE
longhorn-system   default-image-4jjmc   7e8d55ee   download     21430272    117440512     46s
longhorn-system   default-image-kmvkf   006145f3   download     260465664   833617920     66m

# kubectl get vmimages
NAME          DISPLAY-NAME                                       SIZE        VIRTUALSIZE   AGE
image-4jjmc   cirros-0.6.2-x86_64-disk.img                       21430272    117440512     28s
image-kmvkf   opensuse-leap-15.5-minimal-vm.x86_64-cloud.qcow2   260465664                 66m
```

Guess I'll need to fix that...